Objects/unionobject.c

```diff
@@ -137,102 +137,163 @@ union_richcompare(PyObject *a, PyObject *b, int op)
-            total_args += PyTuple_GET_SIZE(((unionobject*) arg)->args);
-        } else {
-            total_args++;
+    int is_ga = _PyGenericAlias_Check(left) && _PyGenericAlias_Check(right);
+    return is_ga ? PyObject_RichCompareBool(left, right, Py_EQ) : left == right;
+}
+
+static int
+is_unionable(PyObject *obj)
+{
+    return (obj == Py_None ||
+        PyType_Check(obj) ||
+        _PyGenericAlias_Check(obj) ||
+        _PyUnion_Check(obj));
+}
+
+static int
+args_contains(PyObject *args, PyObject *obj)
+{
+    assert(PyTuple_CheckExact(args));
+
+    Py_ssize_t size = PyTuple_GET_SIZE(args);
+
+    for (int j = 0; j < size; j++) {
+        PyObject *left_arg = PyTuple_GET_ITEM(args, j);
+        int is_duplicate = is_same(left_arg, obj);
+
+        if (is_duplicate < 0) {
+            return -1;
+        }
+        if (is_duplicate) {
+            return 1;
         }
     }
-    // Create new tuple of flattened args.
-    PyObject *flattened_args = PyTuple_New(total_args);
-    if (flattened_args == NULL) {
+
+    return 0;
+}
+
+static PyObject*
+merge_union_and_union(PyObject *left, PyObject *right)
+{
+    PyObject* left_args = ((unionobject *) left)->args;
+    PyObject* right_args = ((unionobject*) right)->args;
+
+    Py_ssize_t left_size = PyTuple_GET_SIZE(left_args);
+    Py_ssize_t right_size = PyTuple_GET_SIZE(right_args);
+
+    PyObject *tuple = PyTuple_New(left_size + right_size);
+
+    if (tuple == NULL) {
         return NULL;
     }
-    Py_ssize_t pos = 0;
-    for (Py_ssize_t i = 0; i < arg_length; i++) {
-        PyObject *arg = PyTuple_GET_ITEM(args, i);
-        if (_PyUnion_Check(arg)) {
-            PyObject* nested_args = ((unionobject*)arg)->args;
-            Py_ssize_t nested_arg_length = PyTuple_GET_SIZE(nested_args);
-            for (Py_ssize_t j = 0; j < nested_arg_length; j++) {
-                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
-                Py_INCREF(nested_arg);
-                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
-                pos++;
-            }
-        } else {
-            if (arg == Py_None) {
-                arg = (PyObject *)&_PyNone_Type;
-            }
+
+    for (int i = 0; i < left_size; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(left_args, i);
+        Py_INCREF(arg);
+        PyTuple_SET_ITEM(tuple, i, arg);
+    }
+
+    Py_ssize_t pos = left_size;
+
+    for (int i = 0; i < right_size; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(right_args, i);
+
+        int is_duplicate = args_contains(left_args, arg);
+
+        if (is_duplicate < 0) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        if (!is_duplicate) {
             Py_INCREF(arg);
-            PyTuple_SET_ITEM(flattened_args, pos, arg);
+            PyTuple_SET_ITEM(tuple, pos, arg);
             pos++;
         }
     }
-    assert(pos == total_args);
-    return flattened_args;
+
+    if (_PyTuple_Resize(&tuple, pos) < 0) {
```

Comment:
I'm confused about what's this for? Isn't the tuple already the size we want? Or did I miss something important?

Reply:
In case there are duplicates in another union we should not include them in a final union, that why we reduce size of union.

Reply:
Ahh ok thanks that make sense. Could you please insert a one-line comment there explaining that?